### PR TITLE
Bug 1109666 - Customize behavior when long-pressing a link

### PR DESCRIPTION
**Pull request by codestergit**
Implement the long press tap on URL links.If link is valid find the CGRect of URL link in `WKWebView` and we can do lots of things with located links. Currently showing a custom action sheet like safari having options: 
1)Open Link
2)Open link in new tab
3)Copy the link in pasteboard 
4)Save Image
3)Copy Image URL
Please let me know your feedback or suggestions?

### CODE REVIEW

File: Client/Frontend/Browser/Browser.swift [hunk 1]
```diff
@@ -6,10 +6,21 @@ import Foundation
 import WebKit
 
 class Browser {
-    let webView = WKWebView()
+    let webView: WKWebView = {
+
```
Comment (thebnich) [inline Client/Frontend/Browser/Browser.swift]: Nit: Please remove newline

File: Client/Frontend/Browser/Browser.swift [hunk 2]
```diff
@@ -47,4 +58,5 @@ class Browser {
     func loadRequest(request: NSURLRequest) {
         webView.loadRequest(request)
     }
+
```
Comment (thebnich) [inline Client/Frontend/Browser/Browser.swift]: Please remove newline (we generally do not add extra lines before or after curly braces).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 3]
```diff
@@ -0,0 +1,152 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Please fix indentation (the "*"s should line up vertically).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+import Foundation
+import WebKit
+import UIKit
+
+enum ElementType {
+
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: No newline (etc., here and everywhere else below).

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 5]
```diff
@@ -1,6 +1,6 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Please undo this change -- the stars should be aligned :)

File: Client/Frontend/Browser/TabManager.swift [hunk 6]
```diff
@@ -92,7 +95,7 @@ class TabManager {
                 return i
             }
         }
-
+
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: Please remove all trailing whitespace (git rebase HEAD^ --whitespace=fix).

File: Client/Frontend/Browser/TabManager.swift [hunk 7]
```diff
@@ -1,6 +1,6 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: Undo this change

File: Client/JavaScripts/LongPress.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
+function findTagsElementAtPoint(x,y) {
```
Comment (thebnich) [inline Client/JavaScripts/LongPress.js]: All new files need to include the Mozilla license header, so please add that here (copy it from any other Swift file).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    case ImageSrc(url:NSURL,location:CGRect)
+    case hrefLink(url:NSURL,location:CGRect)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Nit: Please add spaces after colons and commas (url: NSURL, location: CGRect). Here and throughout.

File: Client/Frontend/Browser/Browser.swift [hunk 10]
```diff
@@ -6,10 +6,21 @@ import Foundation
...
+        configuration.userContentController = userController
+        return WKWebView(frame: CGRect.zeroRect, configuration: configuration)
+        }()
+
+    let userController: WKUserContentController?
+    let longPressHandler: BrowserLongPressHandler?
 
     init() {
         webView.allowsBackForwardNavigationGestures = true
+        //Added Long Press handler to handle the long press
```
Comment (thebnich) [inline Client/Frontend/Browser/Browser.swift]: You can drop the comment here -- I don't think this code is ambiguous at all.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+            touchLocation.x -= webView!.scrollView.contentInset.left
+            touchLocation.y -= webView!.scrollView.contentInset.top
+            touchLocation.x /= webView!.scrollView.zoomScale
+            touchLocation.y /= webView!.scrollView.zoomScale
+
+            evaluateTouchLocation(touchLocation, gestureRecognizer: gestureRecognizer)
+        }
+    }
+
+    ///Check whether long pressed is on valid link. If found valid link then find the rect for link.
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Nit: Space after "//" or "///" (here and throughout).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+import UIKit
+
+enum ElementType {
+
+    case ImageSrc(url:NSURL,location:CGRect)
+    case hrefLink(url:NSURL,location:CGRect)
+}
+
+protocol BrowserLongPressHandlerDelegate {
+    func webView(webview: WKWebView, elementsDidLongPressed elements: [ElementType])
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Rather than wrapping all of the WebView methods in delegates, we've been simply adding the delegates directly to the WebView in BrowserViewController (<a href="https://github.com/mozilla/firefox-ios/blob/master/Client/Frontend/Browser/BrowserViewController.swift#L99">example</a>). Please use the same pattern here -- remove this delegate, and have BVC assign the delegate directly (tab.webView.addGestureRecognizer(...)).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+class BrowserLongPressHandler: NSObject, UIGestureRecognizerDelegate {
+
+    let webView: WKWebView?
+    var longPressHandlerDelegate: BrowserLongPressHandlerDelegate?
+    let longPressGesture: UILongPressGestureRecognizer?
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Rename longPressGesture -> longPressGestureRecognizer.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+            if tags == nil || tags?.count == 0 {
+                return
+            }
+
+            let selectedTags = tags!.filter({ ($0.asString == "A" || $0.asString == "IMG") })
+            if (selectedTags.count == 0) {
+                return
+            }
+            //
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Remove "//", just have an empty line here.

File: Client/JavaScripts/LongPress.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    return hrefLink;
+}
+
+function findElementsAtPoint(x,y) {
+    var tags = findTagsElementAtPoint(x,y);
+    var jsonResult = {};
+    jsonResult["tags"] = tags
+
+    if(tags.indexOf("A") >= 0) {
```
Comment (thebnich) [inline Client/JavaScripts/LongPress.js]: Nit: space after "if"

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+class BrowserLongPressHandler: NSObject, UIGestureRecognizerDelegate {
+
+    let webView: WKWebView?
+    var longPressHandlerDelegate: BrowserLongPressHandlerDelegate?
+    let longPressGesture: UILongPressGestureRecognizer?
+
+    init(webView: WKWebView) {
+        super.init()
+        self.webView = webView
+        longPressGesture = UILongPressGestureRecognizer(target: self, action: "didLongPress:")
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: didLongPress -> SELdidLongPress (we've been prefixing all selectors with SEL).

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            touchLocation.y -= webView!.scrollView.contentInset.top
+            touchLocation.x /= webView!.scrollView.zoomScale
+            touchLocation.y /= webView!.scrollView.zoomScale
+
+            evaluateTouchLocation(touchLocation, gestureRecognizer: gestureRecognizer)
+        }
+    }
+
+    ///Check whether long pressed is on valid link. If found valid link then find the rect for link.
+    func evaluateTouchLocation(touchLocation: CGPoint,gestureRecognizer: UILongPressGestureRecognizer) {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Make this private (along with all other methods below that aren't called outside of this class).

File: Client/JavaScripts/LongPress.js [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+    if (e && e.href) {
+        hrefLink["hrefLink"] = e.href
+        var r = e.getClientRects()[0];
+        hrefLink["location"] = '{{' + r.left + ',' + r.top + '},{' + r.width + ',' + r.height + '}} ';
+    }
+    return hrefLink;
+}
+
+function findElementsAtPoint(x,y) {
```
Comment (thebnich) [inline Client/JavaScripts/LongPress.js]: Spaces after commas

File: Client/Frontend/Browser/TabManager.swift [hunk 19]
```diff
@@ -52,10 +52,13 @@ class TabManager {
         delegate?.didSelectedTabChange(tab, previous: previous)
     }
 
-    func addTab() -> Browser {
+    func addTab(requestURL: NSURL? = nil) -> Browser {
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: I'd leave this change out of the patch. It might be worth putting in another PR, but it's not related to the issue here.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    case hrefLink(url:NSURL,location:CGRect)
+}
+
+protocol BrowserLongPressHandlerDelegate {
+    func webView(webview: WKWebView, elementsDidLongPressed elements: [ElementType])
+}
+
+class BrowserLongPressHandler: NSObject, UIGestureRecognizerDelegate {
+
+    let webView: WKWebView?
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: This should be non-optional since it's being assigned in init() (and can never become nil after that). Also, please make this variable private since it's not used externally.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+    var longPressHandlerDelegate: BrowserLongPressHandlerDelegate?
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Delegates should be declared using "weak var" to prevent any accidental ref cycles.

File: Client/JavaScripts/LongPress.js [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    var tags = findTagsElementAtPoint(x,y);
```
Comment (thebnich) [inline Client/JavaScripts/LongPress.js]: I don't really see the value of doing this. Why not just unconditionally call findHrefLinkAtPoint/findImageSrcLinkAtPoint?

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+            if (selectedTags.count == 0) {
+                return
+            }
+            //
+            let hrefElement = json["hrefElement"].asDictionary
+            var elements:[ElementType] = []
+            if hrefElement != nil {
+                let hrefStr: String = hrefElement!["hrefLink"]!.asString!
+                let hrefLinkURL = NSURL(string: hrefStr)
+                let linkRect = self.rectLocationInWebView(self.webView!, locationRect:CGRectFromString(hrefElement!["location"]!.asString!))
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Why are you calculating these rects? I don't see them actually used anywhere.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+            if imageElement != nil {
+                let imageSrcStr: String = imageElement!["imageSrc"]!.asString!
+                let imageSrc = NSURL(string: imageSrcStr)
+                let linkRect = self.rectLocationInWebView(self.webView!, locationRect:CGRectFromString(imageElement!["location"]!.asString!))
+                if self.isValidURL(imageSrc) {
+                    elements.append(ElementType.ImageSrc(url: imageSrc!, location: linkRect))
+                }
+            }
+
+            var disableGestures:[UIGestureRecognizer] = []
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: What's going on here? Why are you disabling other gesture recognizers?

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+                            g.enabled = false
+                            disableGestures.append(g)
+                        }
+                    }
+                }
+            }
+
+            //Calls delegate method
+            self.longPressHandlerDelegate?.webView(self.webView!, elementsDidLongPressed: elements)
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(0.5 * Double(NSEC_PER_SEC))), dispatch_get_main_queue()) {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: This looks like an ugly hack. We should definitely look for an alternative approach here.

File: Client/Frontend/Browser/BrowserLongPressHandler.swift [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+                return
+            }
+            //
+            let hrefElement = json["hrefElement"].asDictionary
+            var elements:[ElementType] = []
+            if hrefElement != nil {
+                let hrefStr: String = hrefElement!["hrefLink"]!.asString!
+                let hrefLinkURL = NSURL(string: hrefStr)
+                let linkRect = self.rectLocationInWebView(self.webView!, locationRect:CGRectFromString(hrefElement!["location"]!.asString!))
+                if self.isValidURL(hrefLinkURL) {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: Why only accept "http" and "https" schemes? This may be useful in the Open In New Tab case for schemes the browser can't handle, but I may still want to copy the URL for other schemes, for example. To keep this patch simpler, just omit this and accept all schemes for now.

**Comment by thebnich** [comment]
Thanks for working on this! I'd like to see another iteration here first before granting review. Also, please rebase your updated PR on the latest master branch -- I noticed your commits are based on commits that are nearly a month old that won't be easy to merge.

**Comment by thebnich** [comment]
As you pointed out, <a href="https://github.com/mozilla/firefox-ios/pull/61#discussion_r22818831">this suggestion</a> won't work exactly as described since the gesture recognizer holds a weak reference to the handler, which immediately goes out of scope.

How about making a WebViewLongPressGestureRecognizer class that extends UILongPressGestureRecognizer and implements the UIGestureRecognizerDelegate protocol? Your WebViewLongPressGestureRecognizer init would accept a webView (weak reference), it would call super.init(self, "didLongPress"), and it would set delegate = self. Then your code in BVC would simply be:
        let recognizer = new WebViewLongPressGestureRecognizer(webView: tab.webView)
        tab.webView.addGestureRecognizer(gestureRecognizer: recognizer)

**Comment by thebnich** [comment]
Hey @codestergit, are you still looking at this? Happy to answer any questions here or in IRC, if needed!

File: Client/Frontend/Browser/BrowserLongPressHandler.swift (hunk 24 above)
Comment (codestergit) [inline Client/Frontend/Browser/BrowserLongPressHandler.swift]: @thebnich  I have tried all the things but it seems not possible to avoid default `longPressHandler` with old `document.body.style.webkitTouchCallout='none'` with `WKWebView`.Someone already posted on dev forums https://devforums.apple.com/message/1024729#1024729

**Comment by thebnich** [comment]
Thanks, this is looking a lot better! I'll add my feedback in a moment -- I think this can be merged after one more iteration.

Can you please squash all of your commits into a single commit rebased onto master? We prefer to avoid a bunch of temporary/merge commits to keep the git history as clean as possible. Here's one way to do this easily:
```
git pull origin master
git reset --soft origin/master
git commit -m "Customize behavior when long-pressing a link"
```

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 27]
```diff
@@ -0,0 +1,135 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import WebKit
+import UIKit
+
+enum ElementType {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Since this is going into the global scope, please rename this to LongPressElementType so its purpose is more clear.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+    case ImageSrc(NSURL)
+    case hrefLink(NSURL)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: I don't think we're anticipating any additional elements types, so rather having an array of enums, I think it would be cleaner to use a dictionary of type <code>[LongPressElementType: NSURL]</code> instead. Please remove the NSURL part of these cases and rename them to <code>Image</code> and <code>Link</code> instead.

So, in summary:
```
enum LongPressElementType {
    case Image
    case Link
}
````

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+protocol BrowserLongPressGestureDelegate: class {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: For consistency with the <code>ReaderMode</code> helper, let's remove the Browser prefix and just call this <code>LongPressGestureDelegate</code>.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+    func browser(browser: Browser, elementsDidLongPressed elements: [ElementType])
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: For consistency, please rename <code>browser</code> -> <code>longPressRecognizer</code>, and pass a <code>LongPressGestureRecognizer</code> instead of a <code>Browser</code>. Also, with the dictionary changes above, the <code>elements</code> type should change to a <code>[LongPressElementType: NSURL]</code> dictionary.

So, in summary:
```
func longPressRecognizer(longPressRecognizer: LongPressGestureRecognizer, 
    didLongPressElements elements: [LongPressElementType: NSURL])
```

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+class BrowserLongPressGestureRecognizer: UILongPressGestureRecognizer, UIGestureRecognizerDelegate, BrowserHelper {
+    private  weak var browser: Browser!
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Nit: only one space after <code>private</code>.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+            touchLocation.x /= self.browser.webView.scrollView.zoomScale
+            touchLocation.y /= self.browser.webView.scrollView.zoomScale
+
+            evaluateTouchLocation(touchLocation, gestureRecognizer: gestureRecognizer)
+        }
+    }
+
+    /// Check whether long pressed is on valid link. If found valid link then find the rect for link.
+    private func evaluateTouchLocation(touchLocation: CGPoint,gestureRecognizer: UILongPressGestureRecognizer) {
+        self.browser.webView.evaluateJavaScript("findElementsAtPoint(\(touchLocation.x),\(touchLocation.y))", completionHandler:nil)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Since this is a private one-line function that isn't used anywhere else, you can just inline this into the SELdidLongPress callback above.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 33]
```diff
@@ -0,0 +1,135 @@
...
+        return "BrowserLongPressGestureRecognizer"
+    }
+
+    func scriptMessageHandlerName() -> String? {
+        return "longPressMessageHandler"
+    }
+
+    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        var elementsDict: [String:AnyObject]? = message.body as? [String:AnyObject]
+        if elementsDict  == nil {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Nit: only one space after <code>elementsDict</code>.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+            return
+        }
+
+        var elements:[ElementType] = []
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: With the dictionary changes mentioned above, this will become <code>var elements = \[LongPressElementType: NSURL\]()</code>, and your append calls below will turn into <code>elements[LongPressElementType.Link] = linkURL</code>.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+                        g.enabled = false
+                        disableGestures.append(g)
+                    }
+                }
+            }
+        }
+        if elements.count > 0 {
+            self.browserLongPressGestureDelegate?.browser(browser, elementsDidLongPressed: elements)
+        }
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(0.5 * Double(NSEC_PER_SEC))), dispatch_get_main_queue()) {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: I still strongly dislike doing this, and I'm not even sure you need it. Testing this PR locally, I removed this dispatch and called <code>disableGestures.map({ $0.enabled = true })</code> immediately after calling the delegate above. It still seems to work fine -- so can we remove this delayed dispatch altogether?

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 36]
```diff
@@ -0,0 +1,135 @@
...
+                if let imageSrc = NSURL(string: imageSrcStr) {
+                    elements.append(ElementType.ImageSrc(imageSrc))
+                }
+            }
+        }
+        var disableGestures:[UIGestureRecognizer] = []
+        self.recursiveBlockOnViewAndSubviews(self.browser.webView) { view in
+            if let gestureRecognizers = view.gestureRecognizers as? [UIGestureRecognizer] {
+                for g in gestureRecognizers {
+                    if g != self  && g.enabled == true {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Nit: remove extra space after <code>self</code>.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 37]
```diff
@@ -312,4 +318,56 @@ extension BrowserViewController: ReaderModeDelegate {
             toolbar.updateReaderModeState(state)
         }
     }
+}
+
+extension BrowserViewController: BrowserLongPressGestureDelegate {
+    func browser(browser: Browser, elementsDidLongPressed elements: [ElementType]) {
+        var actionSheetController = UIAlertController(title: "", message: "Mozilla Firefox", preferredStyle: UIAlertControllerStyle.ActionSheet)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Please pass nil as the title instead of an empty string. Also, go ahead and pass nil for message too -- we don't need to show the browser name in this dialog.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 38]
```diff
@@ -312,4 +318,56 @@ extension BrowserViewController: ReaderModeDelegate {
...
+        var url: NSURL?
+        for element in elements {
+            switch element {
+            case .hrefLink(let urlLink):
+                url = urlLink
+                var openNewTabAction =  UIAlertAction(title: "Open In New Tab", style: UIAlertActionStyle.Default) { (action: UIAlertAction!) in
+                    let request =  NSURLRequest(URL:urlLink)
+                    let tab = self.tabManager.addTab(request: request)
+                    return ()  //need explicit void return
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Not needed -- you can remove this line.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 39]
```diff
@@ -312,4 +318,56 @@ extension BrowserViewController: ReaderModeDelegate {
...
+                    var pasteBoard = UIPasteboard.generalPasteboard()
+                    pasteBoard.string = urlLink.absoluteString
+                }
+                actionSheetController.addAction(copyAction)
+
+            default:
+                return
+            }
+        }
+        actionSheetController.title = "\(url!.scheme!)://\(url!.host!)/"
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Please change this to <code>"\(dialogTitleURL!.absoluteString!)"</code> so we show the actual URL of the element being tapped.

File: Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift [hunk 40]
```diff
@@ -0,0 +1,135 @@
...
+                        g.enabled = false
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift]: Note that disabling all gesture recognizers unconditionally will break the default gesture recognizer for other long press actions, namely text selection. Thankfully, the fix for this should be simple: just move this chunk of code (along with the <code>disableGestures.map(...)</code> that follows) into the if condition below. That way, if there are context menu elements, we show the context menu; otherwise, we still allow the default text selection handler to kick in.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 41]
```diff
@@ -312,4 +318,56 @@ extension BrowserViewController: ReaderModeDelegate {
...
     }
+}
+
+extension BrowserViewController: BrowserLongPressGestureDelegate {
+    func browser(browser: Browser, elementsDidLongPressed elements: [ElementType]) {
+        var actionSheetController = UIAlertController(title: "", message: "Mozilla Firefox", preferredStyle: UIAlertControllerStyle.ActionSheet)
+        var url: NSURL?
+        for element in elements {
+            switch element {
+            case .hrefLink(let urlLink):
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: With the dictionary changes mentioned earlier, this function will become a bit nicer. Instead of iterating over the array with the switch statement, you'll need to only do <code>if let linkURL = elements[LongPressElementType.Link] {...}</code> for each.

**Comment by codestergit** [comment]
Addressed the comments.

**Comment by thebnich** [comment]
Looks great, thanks for all the hard work! There is a race condition here with the JS response since we assume we get the response before we disable other delegates, and the other recognizers may be triggered first. But we can handle this in a follow-up and merge this now.
